src/main/java/net/minecraftforge/common/util/PacketUtil.java

```diff
@@ -18,23 +18,65 @@
  */
 package net.minecraftforge.common.util;
 
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.network.PacketBuffer;
+import net.minecraftforge.fml.common.FMLCommonHandler;
+import net.minecraftforge.fml.relauncher.Side;
+
+import javax.annotation.Nullable;
 
 public class PacketUtil
 {
     private PacketUtil() {}
 
+    /**
+     * Serializes an ItemStack to send to the client. If the player is in creative mode, all NBT
+     * data of the item is sent to the client. This prevents the NBT data of the ItemStack being lost
+     * due to the fact that creative serializes from Client to Server, rather than the opposite way.
+     * This still respects the original getShareTag() behavior in PacketBuffer#writeItemStack()
+     *
+     * @param buffer the packet buffer
+     * @param player the player being sent the packet
+     * @param stack the itemstack to serialize
+     */
+    public static void writeItemStackFromServerToClient(PacketBuffer buffer, @Nullable EntityPlayerMP player, ItemStack stack)
```

Comment:
Can you explain the change here?
In the javadoc it says the player is needed for checking creative `client->server` sync, but this method is only intended for `server->client` calls. Can you make a separate method for that case?

Reply:
There is already a method for that in PacketUtil. https://github.com/MinecraftForge/MinecraftForge/pull/4594/files/404fa11a1adea625b464b76fed4b8ace3d4fcbb8#diff-efe8fd11e6036af3738f6bc38b5d936aR79

Reply:
Ok that's good. Can you explain this javadoc and why this method has logic intended for `client->server`?

Reply:
I didn't write the method, I believe you did in https://github.com/MinecraftForge/MinecraftForge/pull/3776. I did though make a slight change to make it send the full NBT (regardless if share tag is false) when in creative mode otherwise this would result in the ItemStack on the server to be updated with an incomplete or blank NBT (Read NetHandlerPlayServer#processCreativeInventoryAction).

Reply:
Sorry I will try to ask my question in a more clear way.

In `writeItemStackFromServerToClient` the javadoc says:

> If the player is in creative mode, all NBT data of the item is sent to the client. This prevents the NBT data of the ItemStack being lost due to the fact that creative serializes from Client to Server, rather than the opposite way.

I do not understand how an itemstack from the server can go to the client and then back to the server when in creative. As far as I understand, `CreativeInventoryAction` is only used for creating new itemstacks, not for moving them around or other actions you described in the original issue. Can you explain this further?

Reply:
I may have confused you with the javadoc, it does not send it back straight away. The only way it sends back is actions as described with `CPacketCreativeInventoryAction`. I was just explaining in the javadoc why we are sending the entire NBT to client regardless if `Item#getShareTag()` is false or `Item#getNBTShareTag()` is returning modified NBT. However now I will explain a scenario to help reason the javadoc further.

As this issue only happens for dedicated servers, when I say client or server I mean physical sides, not logical. It also only happens in creative mode.

First of all, `CPacketCreativeInventoryAction` is sent when a player clicks an inventory slot or when the player drops an item outside the GUI on the client (As described in `PlayerControllerMP`). Straight off the bat we have a problem. Imagine we have an item with NBT in our inventory. That item also overrides `Item#getShareTag()` and returns false. That means the client version of the item will have no NBT as it's not sent because `Item#getShareTag()` is false.

This is where the issue now arises, `CPacketCreativeInventoryAction` sends that item without the NBT to server, and it has now either updated an inventory slot or created an `EntityItem` without the NBT which is bad! When handling `CPacketCreativeInventoryAction` on the server, it uses the `ItemStack` instance deserialized from the packet, not the item instance from the inventory to perform the described actions. We don't use the server's instance of the item in the inventory because creative is `client->server` synchronization.

As stated on the javadoc
> If the player is in creative mode, all NBT data of the item is sent to the client.
From what I wrote in the above paragraphs, handling of `CPacketCreativeInventoryAction` on the server uses the deserialized `ItemStack` from the packet. If we send the full NBT to the client, we get the full NBT back on the server. This again only applies for creative mode only.

Reply:
Thanks for the explanation.
Will there be an issue if there is a similar scenario, when a player gets an item in survival and then goes into creative mode and moves items around in their inventory?

Reply:
I made when you switch gamemode, you are sent an updated version of your inventory. When you switch to creative, you will receive the full NBT.

Reply:
> When you switch to creative, you will receive the full NBT.

May I remind you that the sharetag mechanism wasn't just created to save network traffic but to avoid network packages that are too big to be sent. So this will not work. 

My recommendation would be to change the server-side to try to match the receiver stack to an existing one. To make this easier, Forge could add a UUID to the stacks that are sent to the client either always if they are stripped or when the creative inventory is refreshed.